pkg/metrics_store/metrics_writer.go

```diff
@@ -82,20 +87,59 @@ func (m MetricsWriter) WriteAll(w io.Writer) error {
 	return nil
 }
 
-// SanitizeHeaders removes duplicate headers from the given MetricsWriterList for the same family (generated through CRS).
-// These are expected to be consecutive since G** resolution generates groups of similar metrics with same headers before moving onto the next G** spec in the CRS configuration.
-func SanitizeHeaders(writers MetricsWriterList) MetricsWriterList {
+// SanitizeHeaders sanitizes the headers of the given MetricsWriterList.
+func SanitizeHeaders(contentType string, writers MetricsWriterList) MetricsWriterList {
 	var lastHeader string
 	for _, writer := range writers {
 		if len(writer.stores) > 0 {
 			for i, header := range writer.stores[0].headers {
+
+				// Nothing to sanitize.
+				if len(header) == 0 {
+					continue
+				}
+
+				// Removes duplicate headers from the given MetricsWriterList for the same family (generated through CRS).
+				// These are expected to be consecutive since G** resolution generates groups of similar metrics with same headers before moving onto the next G** spec in the CRS configuration.
+				// Skip this step if we encounter a repeated header, as it will be removed.
+				if header != lastHeader && strings.HasPrefix(header, "# HELP") {
+
+					// If the requested content type was proto-based, replace "info" and "statesets" with "gauge", as they are not recognized by Prometheus' protobuf machinery.
+					if strings.HasPrefix(contentType, expfmt.ProtoType) {
+						infoTypeString := metric.Info.String()
+						stateSetTypeString := metric.StateSet.String()
+						if strings.HasSuffix(header, infoTypeString) {
+							header = header[:len(header)-len(infoTypeString)] + string(metric.Gauge)
+							writer.stores[0].headers[i] = header
```

Comment:
I am not a big fan of modifying the cache here. Ideally, the cache should only be modified when generating metrics after getting an event from an informer.

Also, we shouldn't rely on the assumption that only one format will be served at all time. We never know if a user might have two different backends scraping kube-state-metrics with different formats, so we would just endlessly modify the cache. IMO, it would be better to just remove the type from the cache and generate it here.

Reply:
+1, everything within `ServeHTTP` will be computed for all requests made to it (from one or more metric backends).

Unfortunately, owing the fact that we do not operate on structured metrics in this context, but rather textual representations, I'm not sure if we have a type-independent way of determining the type from the data we receive, other than parsing that text and looking for certain substrings to determine the sanitization behavior.

Quoting https://github.com/kubernetes/kube-state-metrics/pull/2270#discussion_r1432015068,

> +1, but AFAICT we can only deduce the content-type once the request is made (ServeHTTP, and the headers are already generated), which is where SanitizeHeaders is being invoked from currently, for this reason.

> Moving the call further down in the stack, say, to where the headers are being generated wouldn't provide us enough information (content-type) to "sanitize" the headers.

That being said, wouldn't the cache still be modified, since we'd be appending the `# TYPE` metadata to the headers' cache in the case we are able to do so, as compared to editing (`s/typeA/typeB`) it currently?

Reply:
I looked into the code again and I think we could separate the type from the generated headers in the metrics stores: https://github.com/kubernetes/kube-state-metrics/blob/90bd24f27c618ce916db4cd9da34d854d49db5b6/pkg/metrics_store/metrics_store.go#L34-L42.

We could create a new list of types for the metrics families in the metrics store and similarly to how we extract and generate the headers from the metrics families, we could extract the types https://github.com/kubernetes/kube-state-metrics/blob/90bd24f27c618ce916db4cd9da34d854d49db5b6/internal/store/builder.go#L512 and store them in the metrics store.

That will not be a problem when writing the headers because the type is the last element of the headers: https://github.com/kubernetes/kube-state-metrics/blob/90bd24f27c618ce916db4cd9da34d854d49db5b6/pkg/metric_generator/generator.go#L90-L95 so we would just need to append it to the existing headers at runtime.

Reply:
> That being said, wouldn't the cache still be modified, since we'd be appending the # TYPE metadata to the headers' cache in the case we are able to do so, as compared to editing (s/typeA/typeB) it, that is done currently?

Sorry for dragging this on further, but I'm not able to make out how we will circumvent the original concern, i.e., [modifying the cache at runtime](https://github.com/kubernetes/kube-state-metrics/pull/2270#discussion_r1432727642) ([`ServeHTTP`](https://github.com/kubernetes/kube-state-metrics/pull/2270#discussion_r1451795203)), since appending to the header metadata would still fall under that category, and would be similar to what we are doing currently: https://github.com/kubernetes/kube-state-metrics/pull/2270/files#diff-60450a33adea08c953656dd1e78a80e9f3b279bbc7656dedf31fd1a0c7fc1196R108-R109 would effectively become `writer.stores[0].headers[i] += string(metric.Gauge)`, nonetheless, IIUC, we'd still be modifying the header cache at runtime.

I believe we'd still have to deduce the `content-type` at runtime, and modify cache (at runtime) to satisfy the payload expectations. I've provided a comparison of both approaches below.

- [proposed] fetching the metric type from the store and appending the type based on the current `context-type`, in contrast to,
- [current] modifying the last `n` type bytes from the `#TYPE` header based on the current `context-type`.

Although both approaches modify cache at runtime, no fetch operations would be necessary in the latter since we maintain the default header value in the `#TYPE` header (which also translates to less than or equal to the maximum number of appends being done in the previous approach).

PS. PLMK if I missed or misunderstood anything in interpreting your comment above.

Reply:
The type of the metric is not dynamic, it will always be either Gauge, Counter, StateSet or Info. What is dynamic is the type that we will serve to the server depending on the content-type. The one we store doesn't have to change.

My suggestion would be to:
1. remove the type from the header in cache: https://github.com/kubernetes/kube-state-metrics/blob/90bd24f27c618ce916db4cd9da34d854d49db5b6/pkg/metric_generator/generator.go#L93
2. add a slice of types to the metrics store
3. update WriteAll to add the correct type dynamically: https://github.com/kubernetes/kube-state-metrics/blob/main/pkg/metrics_store/metrics_writer.go#L61-L71

for example something like:
```diff
 // MetricsWriterList represent a list of MetricsWriter
@@ -59,6 +63,18 @@ func (m MetricsWriter) WriteAll(w io.Writer) error {
 	}
 
 	for i, help := range m.stores[0].headers {
+
+		if strings.HasPrefix(contentType, expfmt.ProtoType) {
+			switch m.stores[0].types[i] {
+			case metric.Info, metric.StateSet:
+				help += string(metric.Gauge)
+			default:
+				help += string(m.stores[0].types[i])
+			}
+		} else {
+			help += string(m.stores[0].types[i])
+		}
+
 		if help != "" && help != "\n" {
 			help += "\n"
 		}
```

Reply:
Resolved over call, I'll open up a new PR to move types to their own cache, since that should be cleaner.